Call Taker Features + Trip Comparison

landonreed opened the pull request:
This PR adds functionality for Call Taker and basic functionality for RMCE trip comparison (i.e., firing off a batch of trip requests for multiple mode combinations).

There is a handful of stubbed out/partially implemented features, but I figured it would be best to garner feedback at this stage before we went much further. Apologies for the giant PR, but this was the best way I could make progress on these somewhat linked functionalities.

### To test the Call Taker functionality:
1. Start a local instance of otp-datastore.
2. Start with TriMet config `cp ~/git/trimet-mod-otp/config.yml ~/git/otp-react-redux/lib/config.yml`
3. Add the following fields:
4. Visit http://localhost:9966/#/?sessionId=test to avoid auto Trinet redirect

```yaml
datastoreUrl: http://localhost:9000
# datastoreUrl: https://trimet-datastore.ibi-transit.com
trinetReDirect: https://trinet.trimet.org/call_verify_login
```

4. `yarn start`

### To test the RMCE functionality:
1. Start with FDOT RMCE multiple mode combinations config (separate PR).
2. `yarn start`

landonreed commented on lib/components/viewers/pattern-row.js:
```diff
@@ -0,0 +1,210 @@
+import React, { Component } from 'react'
```
Note for reviewer: This component is mostly extracted from stop-viewer to reduce that file's size.

binh-dam-ibigroup commented:
At first browsing, is there any reason for some files to include multiple component classes?

landonreed commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,677 @@
+                  </div>
+                </div>
+              </>
+              : <>
```
Landon to refactor this to be a bit so that FDOT vs. calltaker components are not handled this way.

landonreed commented on lib/util/state.js:
```diff
@@ -25,9 +36,133 @@ export function getActiveItineraries (otpState) {
+/**
+ * This calculates the "cost" (not the monetary cost, but the cost according to
+ * multiple factors like duration, total fare, and walking distance) for a
+ * particular itinerary, for use in sorting itineraries.
```
Big FIXME to actually tune this.

landonreed commented on lib/util/state.js:
```diff
@@ -25,9 +36,133 @@ export function getActiveItineraries (otpState) {
+  const PARKING_COST = 3 * 2.00 * 100
+  // FL per diem rate: https://www.flcourts.org/content/download/219314/1981830/TravelInformation.pdf
+  const PER_MILE_COST_TO_DRIVE = 0.445 * 100
+  const BIKESHARE_FARE = 2 * 100 // $2 per trip? This is a made up number.
```
Config items.

evansiroky replied:
Yes, should be configurable.

binh-dam-ibigroup commented on lib/components/narrative/default/itinerary-summary.js:
```diff
@@ -23,29 +36,48 @@ export default class ItinerarySummary extends Component {
+          title={title}
+          key={blocks.length}
+          className='summary-block mode-block'
+        >
```
Not sure where this observation should go...
The query below results in a missing bikeshare icon:

![image](https://user-images.githubusercontent.com/56846598/88728347-7ff19c80-d0ff-11ea-853a-494112277b4f.png)


http://localhost:9966/#/?ui_activeSearch=pxn1jrumg&ui_activeItinerary=3&fromPlace=600%20Columbia%20St%2C%20Orlando%2C%20FL%2C%20USA%3A%3A28.527074880317347%2C-81.38654708862306&toPlace=2521%20Dade%20Avenue%2C%20Orlando%2C%20FL%2C%20USA%3A%3A28.574572543440226%2C-81.37418746948244&date=2020-07-28&time=18%3A15&arriveBy=false&mode=WALK%2CBUS%2CRAIL&showIntermediateStops=true&maxWalkDistance=1207&optimize=QUICK&walkSpeed=1.34&ignoreRealtimeUpdates=true&companies=&bannedRoutes=&numItineraries=3&otherThanPreferredRoutesPenalty=900&preferredRoutes=

landonreed replied:
Yea, I'm not really sure the source of this issue either. I think the bug is here: 

https://github.com/ibi-group/otp-ui/blob/c8f4f569b3c8a71bd9fd18c4f038a56ca51194b3/packages/icons/src/trimet-mode-icon.js#L28-L33

landonreed replied:
Referenced in https://github.com/opentripplanner/otp-ui/issues/204

binh-dam-ibigroup commented on lib/components/narrative/default/default-itinerary.js:
```diff
@@ -2,50 +2,179 @@ import coreUtils from '@opentripplanner/core-utils'
+                if (bSelected) return 1
+                else return a.order - b.order
+              })
+              .map(attribute => {
```
The bikeshare result for the query below show the "main" activity icon to be "Walk":

![image](https://user-images.githubusercontent.com/56846598/88728556-0312f280-d100-11ea-9b32-d3d10b94339c.png)

http://localhost:9966/#/?ui_activeSearch=pxn1jrumg&ui_activeItinerary=1&fromPlace=600%20Columbia%20St%2C%20Orlando%2C%20FL%2C%20USA%3A%3A28.527074880317347%2C-81.38654708862306&toPlace=2521%20Dade%20Avenue%2C%20Orlando%2C%20FL%2C%20USA%3A%3A28.574572543440226%2C-81.37418746948244&date=2020-07-28&time=18%3A15&arriveBy=false&mode=WALK%2CBUS%2CRAIL&showIntermediateStops=true&maxWalkDistance=1207&optimize=QUICK&walkSpeed=1.34&ignoreRealtimeUpdates=true&companies=&bannedRoutes=&numItineraries=3&otherThanPreferredRoutesPenalty=900&preferredRoutes=

landonreed replied:
I think we should just leave a FIXME for now. There are some unresolved design questions around this. Do we want this to unconditionally show walk time if there is a walking leg? Or should it show combined active time?

binh-dam-ibigroup commented on lib/components/narrative/default/default-itinerary.js:
```diff
@@ -2,50 +2,179 @@ import coreUtils from '@opentripplanner/core-utils'
+              })
+              .map(attribute => {
+                const isSelected = this._isSortingOnAttribute(attribute)
+                const options = attribute.id === 'arrivalTime' ? timeOptions : {}
```
There is an alignment issue for late morning/evening times:

![image](https://user-images.githubusercontent.com/56846598/88728776-7157b500-d100-11ea-8d0f-c473ebc0e936.png)

landonreed replied:
This block is currently set to float right. Not sure the best way to approach a fix for this. Any thoughts?

binh-dam-ibigroup replied:
Use a `min-width` for these blocks?

landonreed replied:
No, doesn't seem to work. This PR is getting stale, so I'm going to move this to an issue.

binh-dam-ibigroup replied:
I posted an alternate layout idea in #211.

binh-dam-ibigroup commented on lib/components/narrative/narrative-itineraries.js:
```diff
@@ -38,19 +70,87 @@ class NarrativeItineraries extends Component {
+              <div>
+                {pending
+                  ? 'Finding your options...'
+                  : <span>We found {itineraries.length} itineraries.</span>
```
On screens that are less than full-width, there are some layout issues:
![image](https://user-images.githubusercontent.com/56846598/88731301-0d83bb00-d105-11ea-8cfb-7b1ec7f2fc45.png)

binh-dam-ibigroup replied:
I do love the layout on full-width, though!

landonreed replied:
Thanks! Yes, we need to resolve mobile layouts as well. I was kind of holding off on spending too much time on the layout concerns until we addressed mobile stuff more fully.

binh-dam-ibigroup commented on lib/components/narrative/default/itinerary.css:
```diff
@@ -4,20 +4,24 @@
-  background-color: var(--hover-color);
+/* Only show border is not expanded */
+.otp .option.default-itin:not(.expanded) {
+  border-bottom: 1px solid grey;
```
I think doing a top border would look better, you would have a separator from the narrative-itineraries (right now there is no line between narrative-itineraries and the first itinerary.

binh-dam-ibigroup replied:
Or add a top border to the first itinerary shown.

landonreed replied:
I had grown used to not having a line between the "We found..." and the first itinerary, but yea perhaps you're right.

binh-dam-ibigroup commented on lib/components/narrative/narrative-itineraries.js:
```diff
@@ -61,17 +161,33 @@ class NarrativeItineraries extends Component {
-          })
-        })}
+        <div
+          className='list'
```
If this is a list with children elements, then this should be an `<ul>`, and the children should be surrounded by `<li>` (with the proper styles).

landonreed replied:
I'm not sure we're semantically that disciplined. But I'm open to being overruled. @evansiroky?

landonreed replied:
Adding fixme for now.

binh-dam-ibigroup commented:
What is this white-ish border between the side pane and the map?
![image](https://user-images.githubusercontent.com/56846598/88737205-e9c17480-d106-11ea-8257-feda65f8fd7f.png)

evansiroky commented on example.js:
```diff
@@ -85,9 +78,15 @@ class OtpRRExample extends Component {
             <Col sm={6} md={4} className='sidebar'>
-              <DefaultMainPanel LegIcon={MyLegIcon} ModeIcon={MyModeIcon} />
+              {/* TODO: extract the BATCH elements out of CallTakerPanel. */}
+              {otpConfig.datastoreUrl || otpConfig.routingTypes.find(t => t.key === 'BATCH')
```
I think I'd prefer for `datastoreUrl` to be named `callTakerDatastoreUrl` so that it indicates that this is used only for the calltaker module.

landonreed replied:
I'm hesitant to do this because this is what it was previously called and it functions for both the call-taker and field-trip modules.

evansiroky commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,677 @@
+    const {expandAdvanced} = this.state
+    return (
+      <ViewerContainer>
+        <div className='main-panel' style={{
+          position: 'absolute',
+          top: 0,
+          left: 0,
+          right: 0,
+          bottom: showPlanTripButton ? 55 : 0,
+          paddingBottom: 15,
+          overflow: 'hidden',
+          padding: '10px'
+        }}>
```
Should we start using styled-components instead of inlining styles like this?

landonreed replied:
Dunno. The only reason I opted for inlining was because I'm not as familiar with styled-components. I'm happy for us to make a decision one way or another.

evansiroky commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,677 @@
+                  flexDirection: 'row',
+                  alignItems: 'center'
+                }} className='comparison-form'>
+                  <button
+                    style={{
+                      height: '50px',
+                      width: '50px',
+                      margin: '0px',
+                      marginRight: '5px'
+                    }}
+                  >
+                    <Icon type='cog' className='fa-2x' />
+                  </button>
+                  <button
+                    style={{
+                      height: '50px',
+                      width: '100px',
+                      margin: '0px',
+                      fontSize: 'small',
+                      textAlign: 'left'
+                    }}
```
These 2 buttons don't do anything when clicked.

evansiroky replied:
Why is this resolved? These buttons still don't do anything when clicked.

evansiroky commented on lib/actions/api.js:
```diff
@@ -83,86 +82,96 @@ function getActiveItinerary (otpState) {
+    // one iteration.
+    const iterations = otpState.config.modes && otpState.config.modes.combinations
+      ? otpState.config.modes.combinations.map(({mode, params}) => ({mode, ...params}))
+      : [{}]
```
Won't this result in no trip plans happening with the current trimet-call-taker config?

landonreed replied:
No, it should result in one, the empty object represents the trip plan -- just with no injected params.

evansiroky commented on lib/actions/api.js:
```diff
@@ -83,86 +82,96 @@ function getActiveItinerary (otpState) {
+        user.loggedInUser &&
+        user.loggedInUser.storeTripHistory
+
+      fetch(constructRoutingQuery(otpState, true), getOtpFetchOptions(state, storeTripHistory))
+        .then(getJsonAndCheckResponse)
+        .then(json => {
+          // FIXME: This is only performed when ignoring realtimeupdates currently, just
+          // to ensure it is not repeated twice.
+          dispatch(nonRealtimeRoutingResponse({ response: json, searchId }))
+        })
```
Since we're now making queries for all kinds of mode combinations, we should probably check some of the mode combinations that won't have realtime data and avoid making requests for those types of queries.

landonreed replied:
Will add FIXME for now.

evansiroky commented on lib/actions/call-taker.js:
```diff
@@ -0,0 +1,209 @@
+        // Inject call ID into active call and save queries.
+        dispatch(saveQueriesForCall({...activeCall, id}))
+          .then(() => {
+            console.log('finished saving queries for call', id)
```
remove console.log

evansiroky commented on lib/components/form/intermediate-place-field.js:
```diff
@@ -0,0 +1,112 @@
+  render () {
+    return (
+      <>
+        <StyledIntermediatePlace {...this.props} />
```
This intermediate place needs a different marker than the from/to icons.

evansiroky commented on lib/components/form/intermediate-place-field.js:
```diff
@@ -0,0 +1,112 @@
+    return (
+      <>
+        <StyledIntermediatePlace {...this.props} />
+        <button
+          className='pull-right clear-button-formatting'
+          style={{
+            position: 'relative',
+            top: '-50px',
+            left: '-13px'
+          }}
+          onClick={this._removePlace}>
+          x
+        </button>
```
This button overlaps the location-field clear button.

<img width="431" alt="Screen Shot 2020-07-29 at 5 45 26 PM" src="https://user-images.githubusercontent.com/3112493/88867873-8a756a00-d1c3-11ea-86dc-428c921c0686.png">

landonreed replied:
Fixed in 44d4a00

evansiroky commented on lib/components/map/intermediate-places-overlay.js:
```diff
@@ -0,0 +1,51 @@
+      <>
+        {Array.isArray(intermediatePlaces) && intermediatePlaces.map((place, i) => {
+          return place && place.lat && place.lon
+            ? <Marker key={i} position={[place.lat, place.lon]} />
```
Whatever icon that is used for this marker should match up to the marker in the location-field and also the one shown in the itinerary result.

evansiroky commented on lib/components/narrative/default/itinerary-summary.js:
```diff
@@ -1,6 +1,19 @@
 import PropTypes from 'prop-types'
 
+function getModeColor (mode) {
+  switch (mode) {
+    case 'BICYCLE': return '#E0C3E2'
+    case 'BUS': return '#CAC3DF'
+    case 'CAR': return '#E4CCCC'
+    // case 'PARK': return '#E4CCCC'
+    case 'RAIL': return '#BDDAC0'
+    case 'WALK': return '#DFC486'
+    default:
+      return 'grey'
+  }
```
It'd probably be better to use a hashmap here so that each case statement doesn't need to be analyzed.

binh-dam-ibigroup commented on lib/components/admin/call-taker-windows.js:
```diff
@@ -0,0 +1,100 @@
+                position: 'absolute',
+                zIndex: 9999999,
+                width: '350px',
+                height: '350px',
```
Add overflow handling (see below)

![image](https://user-images.githubusercontent.com/56846598/88932771-83466e80-d24c-11ea-9994-8f1f95fc3bfa.png)

binh-dam-ibigroup commented on lib/util/state.js:
```diff
@@ -12,6 +15,14 @@ export function getActiveSearch (otpState) {
+ * Get timestamp in the expected format used by otp-datastore (call taker
+ * back end). Defaults to now.
+ */
+export function getTimestamp (time = moment()) {
```
~Rename to `getISOTimeString` Also~ add TODO to potentially move to OTP-UI.

evansiroky commented on lib/components/admin/call-taker-controls.js:
```diff
@@ -0,0 +1,157 @@
+      console.warn('No active call is in progress! Cannot get call time.')
+      return ''
+    }
+    console.log(moment().diff(moment(this.props.activeCall.startTime)))
```
Remove console.log

landonreed commented on lib/components/narrative/narrative-itineraries.js:
```diff
@@ -61,17 +161,33 @@ class NarrativeItineraries extends Component {
+            const active = index === activeItinerary
+            // Hide non-active itineraries.
+            if (!active && itineraryIsExpanded) return null
+            return React.createElement(itineraryClass, {
```
Check out the difference between https://newplanner.trimet.org and https://trimet-mod-dev.ibi-transit.com/ . This is an issue with otp-ui or its implementation in otp-react-redux.

### https://trimet-mod-dev.ibi-transit.com
![image](https://user-images.githubusercontent.com/2370911/88968371-3e86fb80-d27d-11ea-99d0-eeaf013e3eba.png)

### https://trimet-mod-dev.ibi-transit.com/
![image](https://user-images.githubusercontent.com/2370911/88968410-4cd51780-d27d-11ea-8ba3-fb4024f20ca3.png)

landonreed replied:
Moved to issue: https://github.com/opentripplanner/otp-react-redux/issues/212

landonreed commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,677 @@
+
+class DateTimeOptions extends Component {
+  _setDepartArrive = evt => {
+    const {value: departArrive} = evt.target
```
I think this might be related to an error being thrown in calltaker that is new as of some recent changes in otp-ui. I think this will help: https://github.com/opentripplanner/otp-ui/pull/193

landonreed commented on lib/actions/api.js:
```diff
@@ -217,13 +226,16 @@ function constructRoutingQuery (otpState, ignoreRealtimeUpdates) {
+  const params = getRoutingParams(currentQuery, ignoreRealtimeUpdates)
+  // Apply mode override, if specified (for batch routing).
+  Object.keys(injectedParams).forEach(key => {
+    params[key] = injectedParams[key]
```
Yea, not sure why I opted for that. I was probably copying code from elsewhere and wasn't thinking.

landonreed commented on lib/components/admin/call-record.js:
```diff
@@ -0,0 +1,116 @@
+              className='animate-flicker' />
+            <CallTimeCounter
+              style={{display: 'inline'}}
+              // FIXME For some reason this counter is not updating...
```
fixed.

landonreed commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,677 @@
+    return bannedRoutes && bannedRoutes.find(o => o.value === option.value)
+  }
+
+  getDistanceStep = distanceInMeters => {
```
This isn't really for humanizing distance, but this input might no longer use this method at some point anyways.

landonreed commented on lib/components/map/intermediate-places-overlay.js:
```diff
@@ -0,0 +1,51 @@
+} from '../../actions/map'
+import { getShowUserSettings } from '../../util/state'
+
+class IntermediatePlacesOverlay extends Component {
```
Done.

landonreed commented on lib/actions/call-taker.js:
```diff
@@ -0,0 +1,209 @@
+      formData.append('query.queryParams', JSON.stringify(queryParams))
+      formData.append('query.fromPlace', from.name || placeToLatLonStr(from))
+      formData.append('query.toPlace', to.name || placeToLatLonStr(to))
+      formData.append('query.timeStamp', search.query.timestamp)
```
What would you suggest we do if the timestamp is null in that case? Simply skip persisting the timestamp? It seems like something that should maybe be handled in the backend for now.

landonreed commented on lib/components/admin/call-record.js:
```diff
@@ -0,0 +1,116 @@
+        <button onClick={this._viewQuery} className='clear-button-formatting'>
+          <Icon type='search' />
+        </button>{' '}
+        Query {index + 1}
```
This whole layout is subject to change, so I'll skip for now.

evansiroky commented on lib/components/app/call-taker-panel.js:
```diff
@@ -586,7 +561,12 @@ class DateTimeOptions extends Component {
   handleTimeChange = evt => {
-    this.props.setQueryParam({ time: evt.target.value })
+    const timeInput = evt.target.value
+    console.log(timeInput)
+    const date = moment().startOf('day').format('YYYY-MM-DD')
+    const time = moment(date + 'T' + timeInput, TIME_FORMATS)
+    console.log(time, time.format(OTP_API_TIME_FORMAT))
```
Remove console.logs

evansiroky commented on lib/components/form/connected-settings-selector-panel.js:
```diff
@@ -23,7 +23,7 @@ class ConnectedSettingsSelectorPanel extends Component {
       showUserSettings
     } = this.props
-
+    console.log(config.modes, config.companies)
```
Remove

evansiroky commented:
The walkTime LegIcon walks into the itinerary.


![Screen Recording 2020-08-17 at 6 11 57 PM](https://user-images.githubusercontent.com/3112493/90459405-1a694e00-e0b6-11ea-9fd9-00a5aa24e95f.gif)

evansiroky commented:
The route grey circles should have the same background color as the preceding mode icon background color.

<img width="192" alt="Screen Shot 2020-08-17 at 6 19 46 PM" src="https://user-images.githubusercontent.com/3112493/90459499-5b616280-e0b6-11ea-810b-dcdbdd1ef52a.png">

evansiroky commented on lib/util/state.js:
```diff
@@ -25,9 +37,134 @@ export function getActiveItineraries (otpState) {
+    itinerary.duration * DURATION_COST_FACTOR +
+    itinerary.walkDistance * WALK_COST_FACTOR +
+    itinerary.waitingTime * WAIT_COST_FACTOR +
+    itinerary.transfers * TRANSFER_COST_FACTOR
```
I think it'd be useful to include a drive reluctance factor in here. I think most of our customers will be transit agencies and will therefore not want itineraries with lots of driving to show up so high in the results.

evansiroky commented on lib/util/state.js:
```diff
@@ -25,9 +37,134 @@ export function getActiveItineraries (otpState) {
+ * FIXME: Do major testing to get this right.
+ */
+function calculateItineraryCost (itinerary) {
+  const FARE_COST_FACTOR = 0.5
+  const DURATION_COST_FACTOR = 0.25
+  const WALK_COST_FACTOR = 0.1
+  const WAIT_COST_FACTOR = 0.1
+  const TRANSFER_COST_FACTOR = 0.9
```
These should all be configurable.

evansiroky commented on lib/components/narrative/narrative-itineraries.js:
```diff
@@ -30,48 +33,174 @@ class NarrativeItineraries extends Component {
+    this.setState({showDetails: !this.state.showDetails})
+  }
+
+  _saveTrip = () => {
+    window.confirm('Are you sure you want to save this trip?')
+  }
```
Doesn't save trip, just displays confirmation window.

landonreed replied:
This was built before @binh-dam-ibigroup's work existed. I think we should coordinate on syncing this work outside of this PR.

evansiroky commented:
There ought to be a change to the bounds-updating-overlay to make sure changes to intermediate places will be fit inside new bounds.

https://github.com/opentripplanner/otp-react-redux/blob/a6b1710190b15499bd8b8fde9fc2d2916d0f8b52/lib/components/map/bounds-updating-overlay.js#L33

evansiroky commented:
Itineraries are hidden behind main panel in call taker panel.

<img width="426" alt="Screen Shot 2020-08-18 at 12 14 56 AM" src="https://user-images.githubusercontent.com/3112493/90482305-e149d180-e0e7-11ea-910f-afbbd469cbca.png">

evansiroky commented on lib/components/map/connected-endpoints-overlay.js:
```diff
@@ -17,9 +17,13 @@ const mapStateToProps = (state, ownProps) => {
   const showUserSettings = getShowUserSettings(state.otp)
-  const { from, to } = query
+  const { from, intermediatePlaces, to } = query
+  // FIXME: Splitting places by comma is broken in otp-ui:
+  // https://github.com/opentripplanner/otp-ui/issues/203
```
This should be blocking as it is making it impossible to plan trips with 2+ intermediate places.

landonreed replied:
It's breaking reloading an itinerary from a URL as well, but we don't have a good fix right now and it's not affecting our production code, so I don't think it should be blocking for this MVP at the moment.

evansiroky commented on lib/components/narrative/narrative-itineraries.js:
```diff
@@ -30,48 +33,174 @@ class NarrativeItineraries extends Component {
     } = this.props
-    if (!itineraries) return null
-
+    if (!activeSearch) return null
+    const itineraryIsExpanded = activeItinerary !== undefined && activeItinerary !== null && this.state.showDetails
+    const showRealtimeAnnotation = realtimeEffects.isAffectedByRealtimeData && (
+      realtimeEffects.exceedsThreshold ||
+      realtimeEffects.routesDiffer ||
+      !useRealtime
+    )
+    const resultText = pending
+      ? 'Finding your options...'
+      : `${itineraries.length} itineraries found.`
```
Doesn't ever show an error message if one is returned in OTP response.

See usage in narrative-routing-results

https://github.com/opentripplanner/otp-react-redux/blob/a6b1710190b15499bd8b8fde9fc2d2916d0f8b52/lib/components/narrative/narrative-routing-results.js#L30

landonreed replied:
fixed.

evansiroky commented:
I really want to drag these things around, but I can't. Not sure if feature or bug.

<img width="177" alt="Screen Shot 2020-08-18 at 12 33 31 AM" src="https://user-images.githubusercontent.com/3112493/90484062-85cd1300-e0ea-11ea-82f9-cf1c1257a887.png">

binh-dam-ibigroup commented:
> The walkTime LegIcon walks into the itinerary.
> 
> ![Screen Recording 2020-08-17 at 6 11 57 PM](https://user-images.githubusercontent.com/3112493/90459405-1a694e00-e0b6-11ea-9fd9-00a5aa24e95f.gif)

This also appears in the itinerary summaries:
![Screen Recording 2020-08-18 at 10 41 13 AM](https://user-images.githubusercontent.com/56846598/90529931-43cfbb80-e142-11ea-971b-5b9f659d6c42.gif)

binh-dam-ibigroup commented:
> Itineraries are hidden behind main panel in call taker panel.
> 
> <img alt="Screen Shot 2020-08-18 at 12 14 56 AM" width="426" src="https://user-images.githubusercontent.com/3112493/90482305-e149d180-e0e7-11ea-910f-afbbd469cbca.png">

The bottom border might need a shadow effect similar to the one added in this PR to the main pane.

binh-dam-ibigroup commented on lib/components/admin/query-record.js:
```diff
@@ -0,0 +1,42 @@
+        <button onClick={this._viewQuery} className='clear-button-formatting'>
+          <Icon type='search' />
+        </button>{' '}
+        Query {index + 1}
```
This should be part of the button IMO.

landonreed replied:
I think this was already discussed in a previous conversation. This whole component is subject to change.

binh-dam-ibigroup commented:
Between 10 and 12 am/pm, the leg departure/arrival time breaks into two lines, I don't think that is intended.
![image](https://user-images.githubusercontent.com/56846598/90543567-e5abd400-e153-11ea-88fc-345f03fc9dd4.png)

binh-dam-ibigroup commented on lib/components/narrative/default/itinerary-summary.js:
```diff
@@ -23,29 +40,48 @@ export default class ItinerarySummary extends Component {
-      // If a transit leg, add the name (preferably short; long if needed)
-      if (leg.transitLeg) {
+      // If a transit leg, add the name (only if short name exists)
+      if (leg.transitLeg && leg.routeShortName) {
```
This should not repeat a non-transit mode IMO:
 
![image](https://user-images.githubusercontent.com/56846598/90543932-5fdc5880-e154-11ea-9082-6d5150ede694.png)

Query: http://localhost:9966/#/?ui_activeSearch=gzw0a0kme&ui_activeItinerary=0&fromPlace=1222%20SW%20Davenport%20St%2C%20Portland%3A%3A45.505143833370234%2C-122.69187927246095&toPlace=3504%20NE%20Us%20Grant%20Pl%2C%20Portland%3A%3A45.53797835175697%2C-122.62819290161134&date=2020-08-18&time=11%3A35&arriveBy=false&mode=WALK%2CBUS%2CTRAM%2CRAIL%2CGONDOLA&showIntermediateStops=true&maxWalkDistance=1207&optimize=QUICK&walkSpeed=1.34&ignoreRealtimeUpdates=true&companies=&bannedRoutes=&numItineraries=3&intermediatePlaces=Rose%20Quarter%20TC%20MAX%20Station%2C%20Portland%3A%3A45.530147%2C-122.665225&otherThanPreferredRoutesPenalty=900&preferredRoutes=

landonreed replied:
This trip is for an intermediate place, so there are two distinct legs in the itinerary for walking. I somewhat disagree, but in general I think this trip summary component is not quite finished and needs refinement, which I think we should handle separately.

binh-dam-ibigroup replied:
Ooh ok. In that case the summary band would need to show an intermediate place indicator/separator.

landonreed commented on lib/components/app/batch-routing-panel.js:
```diff
@@ -0,0 +1,120 @@
+              marginRight: '5px'
+            }}
+          >
+            <Icon type='cog' className='fa-2x' />
```
It's stubbed out.

landonreed commented on lib/components/app/batch-routing-panel.js:
```diff
@@ -0,0 +1,120 @@
+              textAlign: 'left'
+            }}
+          >
+            <Icon type='calendar' /> Today<br />
```
It's stubbed out.

landonreed commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,652 @@
+          placeholder='Select banned routes...' />
+        <Select
+          value={preferredRoutes}
+          id='preferredRoutes'
```
Sure. Will change.

landonreed commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,652 @@
+            display: 'inline-flex'
+          }}>
+            <span>{cleanTime}</span>
+            <input
```
I'm going to leave as is for now. This needs user testing from TriMet and is really just a placeholder input for now.

landonreed commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,652 @@
+                onBlur={this._setMode}
+                onKeyDown={this._handleFormKeyDown}
+                value={this.modeToOptionValue(mode)}
+                style={{position: 'absolute', right: '50px'}}
```
I anticipate the form style will need to be reworked. This is basically just an MVP.

landonreed commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,652 @@
+          }
+          <StyledSubmodeSelector
+            modes={transitModes}
+            onChange={this._onSubModeChange}
```
fixed.

landonreed commented on example.css:
```diff
@@ -34,6 +34,8 @@
   height: 100%;
   padding: 10px;
   overflow-y: scroll;
```
I think I accidentally changed a class name that is meant to have a scroll bar. Next commit will fix this.

landonreed commented on lib/components/app/batch-routing-panel.js:
```diff
@@ -0,0 +1,120 @@
+              display: 'flex',
+              flexDirection: 'column',
+              position: 'absolute',
+              top: '218px', // This is variable dependent on height of the form above.
```
Could you clarify your comment?

binh-dam-ibigroup replied:
The thing is if you set a fixed top, there will be a good deal of blank space between trip options and the itineraries when the advanced options pane is collapsed. That can be dealt with later.

landonreed commented on lib/components/app/call-taker-panel.js:
```diff
@@ -0,0 +1,652 @@
+              <i className='fa fa-plus-circle' /> Add place
+            </button>
+            <div className='search-options' style={{height: '30px'}}>
+              <DateTimeOptions
```
I'm not entirely sure what you mean, but I think we should refine styling later on.

binh-dam-ibigroup commented:
> Since there are so many unresolved issues, it seems unwise to approve merging this to dev, so I'm just going to make a comment. I think about 4 of the 6 items mentioned in #216 are show-stoppers that ought to be addressed. Btw, I made a [proposed fix](https://github.com/opentripplanner/otp-ui/pull/210) in OTP UI with respect to the intermediate places.

Am I the tie breaker here?

landonreed commented:
I don't think blocking future releases from `dev` is a reasonable condition, so I suppose I'll just fix these items now.

landonreed commented:
@evansiroky, I've addressed 5 of 6 items in #216. The one remaining relates to a stubbed out feature that hasn't been fully designed yet, so I don't think it's wise to include it here. I should also note that I incorporated your fix for the intermediate place parsing, but it doesn't seem to work with the OTP API for some reason.

evansiroky commented on lib/components/app/call-taker-panel.js:
```diff
@@ -285,8 +289,13 @@ class CallTakerPanel extends Component {
-              <i className='fa fa-plus-circle' /> Add place
+              <i className='fa fa-plus-circle' />{' '}
+              {addIntermediateDisabled
+                ? 'Define origin/destination to add intermediate places'
```
This message will display also when the max number of intermediate places has been reached.

evansiroky commented on lib/util/state.js:
```diff
@@ -106,75 +106,116 @@ export function getActiveItineraries (otpState) {
+const WALK_RELUCTANCE = 0.1
+const WAIT_RELUCTANCE = 0.1
+const DRIVE_RELUCTANCE = 2
+const TRANSFER_RELUCTANCE = 0.9
```
These can be moved to the constants file.

evansiroky commented on lib/util/state.js:
```diff
@@ -106,75 +106,116 @@ export function getActiveItineraries (otpState) {
-    itinerary.waitingTime * WAIT_COST_FACTOR +
-    itinerary.transfers * TRANSFER_COST_FACTOR
+function calculateItineraryCost (itinerary, config = {}) {
+  const weights = {
```
Can you define a `defaultWeights` variable instead of writing out each variable individually?

evansiroky commented on lib/util/state.js:
```diff
@@ -106,75 +106,116 @@ export function getActiveItineraries (otpState) {
-  const PER_MILE_COST_TO_DRIVE = 0.445 * 100
-  const BIKESHARE_FARE = 2 * 100 // $2 per trip? This is a made up number.
+  // Start with default weighting values.
+  const costs = {
+    bikeshareTripCostCents: DEFAULT_BIKESHARE_FARE,
+    carParkingCostCents: DEFAULT_PARKING_COST,
+    drivingCentsPerMile: DEFAULT_DRIVING_COST_PER_MILE
+  }
```
Can you define a `defaultCosts` variable instead of writing out each variable individually?

evansiroky commented on lib/util/state.js:
```diff
@@ -106,75 +106,116 @@ export function getActiveItineraries (otpState) {
+// $2 for 3 hours of parking?
+const DEFAULT_PARKING_COST = 3 * 2.00 * 100
+// FL per diem rate: https://www.flcourts.org/content/download/219314/1981830/TravelInformation.pdf
+const DEFAULT_DRIVING_COST_PER_MILE = 0.445 * 100
```
These could go in the constants file

landonreed replied:
If they are not used anywhere but this file, is there really any benefit to moving them there?

evansiroky reviewed (changes requested):
I pushed a change that fixes the intermediate places stuff. Three things I noticed that would be nice-to-haves with intermediate places:
- ability to drag the intermediate place map markers
- ability to drag/drop the intermediate places to reorder them in the form
- the reverse from/to button should probably just be removed if there is at least one intermediate place

Some things to fix still:

- Travis Build is failing
- messaging about too many intermediate places
- some refactoring ideas with factors/costs (this is optional I suppose)

evansiroky reviewed (approved):
Those final 3 fixes I requested are fixed, so I'm apprvoing.

landonreed commented:
:tada: This PR is included in version 1.3.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/otp-react-redux)
- [GitHub release](https://github.com/opentripplanner/otp-react-redux/releases/tag/v1.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
